### ISSUE TITLE
[ARM/Linux] Unittest failure 'Interop/SizeConst/SizeConstTest'

### ISSUE THREAD

**Issue by parjong** [dotnet/coreclr#6569]
In ARM/Linux, CoreCLR shows the following assertion failure while running `Interop/SizeConst/SizeConstTest`:

```
Assert failure(PID 16645 [0x00004105], Thread: 16645 [0x4105]): !IsTypeDesc()
    File: /home/parjong/projects/toolchains/coreclr/src/vm/typehandle.inl Line: 115
    Image: /opt/dotnet/Linux.tm1.Debug.Debug/corerun


Program received signal SIGABRT, Aborted.
0xb6d01c64 in raise () from /lib/libc.so.6
(gdb) bt
#0  0xb6d01c64 in raise () from /lib/libc.so.6
#1  0xb6d02ff8 in abort () from /lib/libc.so.6
#2  0xb6aaa52a in PROCEndProcess (hProcess=0xffffff01, uExitCode=1, bTerminateUnconditionally=1) at /home/parjong/projects/toolchains/coreclr/src/pal/src/thread/process.cpp:1385
#3  0xb6aaa67e in TerminateProcess (hProcess=0xffffff01, uExitCode=1) at /home/parjong/projects/toolchains/coreclr/src/pal/src/thread/process.cpp:1301
#4  0xb63d24dc in _DbgBreakCheck (szFile=0xb6ad02ac "/home/parjong/projects/toolchains/coreclr/src/vm/typehandle.inl", iLine=115, szExpr=0xb6ad02ec "!IsTypeDesc()", fConstrained=0)
    at /home/parjong/projects/toolchains/coreclr/src/utilcode/debug.cpp:492
#5  0xb63d27c4 in _DbgBreakCheckNoThrow (szFile=0xb6ad02ac "/home/parjong/projects/toolchains/coreclr/src/vm/typehandle.inl", iLine=115, szExpr=0xb6ad02ec "!IsTypeDesc()", fConstrained=0)
    at /home/parjong/projects/toolchains/coreclr/src/utilcode/debug.cpp:558
#6  0xb63d2b22 in DbgAssertDialog (szFile=0xb6ad02ac "/home/parjong/projects/toolchains/coreclr/src/vm/typehandle.inl", iLine=115, szExpr=0xb6ad02ec "!IsTypeDesc()")
    at /home/parjong/projects/toolchains/coreclr/src/utilcode/debug.cpp:745
#7  0xb63f9536 in TypeHandle::AsMethodTable (this=0xbeffb3c0) at /home/parjong/projects/toolchains/coreclr/src/vm/typehandle.inl:115
#8  0xb642e348 in CEEInfo::getFieldInClass (this=0xbeffd530, clsHnd=0xb038447e, num=0) at /home/parjong/projects/toolchains/coreclr/src/vm/jitinterface.cpp:2220
#9  0xaffe1646 in Compiler::isSingleFloat32Struct (this=0x9cb98, clsHnd=0xb038447e) at /home/parjong/projects/toolchains/coreclr/src/jit/compiler.cpp:475
#10 0xaffe1762 in Compiler::getPrimitiveTypeForStruct (this=0x9cb98, structSize=4, clsHnd=0xb038447e) at /home/parjong/projects/toolchains/coreclr/src/jit/compiler.cpp:572
#11 0xaffe18da in Compiler::getArgTypeForStruct (this=0x9cb98, clsHnd=0xb038447e, wbPassStruct=0xbeffb640, structSize=4) at /home/parjong/projects/toolchains/coreclr/src/jit/compiler.cpp:730
#12 0xb00b29a2 in Compiler::fgMorphArgs (this=0x9cb98, callNode=0xa0b60) at /home/parjong/projects/toolchains/coreclr/src/jit/morph.cpp:3405
#13 0xb00c466c in Compiler::fgMorphCall (this=0x9cb98, call=0xa0b60) at /home/parjong/projects/toolchains/coreclr/src/jit/morph.cpp:7958
#14 0xb00b1ad0 in Compiler::fgMorphTree (this=0x9cb98, tree=0xa0b60, mac=0x0) at /home/parjong/projects/toolchains/coreclr/src/jit/morph.cpp:13848
#15 0xb00bd750 in Compiler::fgMorphSmpOp (this=0x9cb98, tree=0xa0e5c, mac=0x0) at /home/parjong/projects/toolchains/coreclr/src/jit/morph.cpp:10544
#16 0xb00b1a6a in Compiler::fgMorphTree (this=0x9cb98, tree=0xa0e5c, mac=0x0) at /home/parjong/projects/toolchains/coreclr/src/jit/morph.cpp:13835
#17 0xb00c820e in Compiler::fgMorphStmts (this=0x9cb98, block=0x9f6ac, mult=0xbeffc957, lnot=0xbeffc956, loadw=0xbeffc955) at /home/parjong/projects/toolchains/coreclr/src/jit/morph.cpp:14599
#18 0xb00c89e2 in Compiler::fgMorphBlocks (this=0x9cb98) at /home/parjong/projects/toolchains/coreclr/src/jit/morph.cpp:14868
#19 0xb00ca8e2 in Compiler::fgMorph (this=0x9cb98) at /home/parjong/projects/toolchains/coreclr/src/jit/morph.cpp:15969
```

**Comment by parjong** [comment]
PR #6679 eliminates the above assertion failure, but introduces the following uncaught exception:
```
Unhandled Exception: System.Runtime.InteropServices.COMException: The data area passed to a system call is too small.

 (Exception from HRESULT: 0x8007007A)

   at System.StubHelpers.ValueClassMarshaler.ConvertToNative(IntPtr dst, IntPtr src, IntPtr pMT, CleanupWorkList& pCleanupWorkList)

   at Test.TakeByValTStr(S_CHARArray_ByValTStr s, Int32 size)

   at Test.SizeConstByValTStr()

   at Test.Main(String[] args)
```

**Comment by mazong1123** [comment]
I can reproduce this issue on Windows 7 x64 with exactly the same exception message:

```
Unhandled Exception: System.Runtime.InteropServices.COMException: The data area passed to a system call is too small.
 (Exception from HRESULT: 0x8007007A)
   at System.StubHelpers.ValueClassMarshaler.ConvertToNative(IntPtr dst, IntPtr src, IntPtr pMT, CleanupWorkList& pCleanupWorkList)
   at Test.TakeByValTStr(S_CHARArray_ByValTStr s, Int32 size)
   at Test.SizeConstByValTStr()
   at Test.Main(String[] args)
```

**Comment by parjong** [comment]
@mazong1123 The issue seems to be related with non-ASCII encoding.

**Comment by mazong1123** [comment]
@parjong thanks! Do you know is this related to my locale setting? I'm also curious is this by design or a bug.

**Comment by parjong** [comment]
@mazong1123 I'm not sure about that. I'm working on another issue as 8d51b631b3f commit disables this test. 

I just guess that this issue is related with some encoding because Core CLR internally allocates 4 bytes for P/Invoke interop, but the strict itself requires 8 bytes (although it consists of 4 characters).

**Comment by jkoritzinsky** [comment]
Fixed by #20268
